lib/membrane_audio_mixer/do_interleave.ex

```diff
@@ -0,0 +1,74 @@
+defmodule Membrane.AudioMixer.DoInterleave do
+  @moduledoc """
+  Module responsible for interleaving audio tracks (all in the same format, with 1
+  channel) in a given order.
+  """
+
+  @doc """
+  Order queues according to `order`, take `bytes_per_channel` from each queue
+  (all queues must be at least `bytes_per_channel` long),
+  and interleave them.
+  """
+  @spec interleave(pos_integer(), pos_integer(), %{}, [integer()]) :: {any(), map()}
+  def interleave(bytes_per_channel, _sample_size, pads, _order) when map_size(pads) == 1 do
+    [{pad, data}] = Map.to_list(pads)
+
+    <<payload::binary-size(bytes_per_channel)>> <> remaining_queue = data.queue
+    pads = %{pad => %{data | queue: remaining_queue}}
+
+    {payload, pads}
+  end
+
+  def interleave(bytes_per_channel, sample_size, pads, order) do
+    pads_inorder = order_pads(pads, order)
+    {payloads, pads_list} = get_payloads(bytes_per_channel, pads_inorder)
+
+    payload = interleave_binaries(payloads, sample_size)
+
+    {payload, Map.new(pads_list)}
+  end
+
+  #  Interleave binaries, taking `sample_size` bytes at a time.
+  defp interleave_binaries(payloads, sample_size) do
+    payloads
+    # split each channel's payload into `sample_size` chunks (channels order is reversed)
+    |> Enum.map(fn payload -> to_chunks_reversed(payload, sample_size) end)
+    # zip corresponding chunks of different channels and concatenate them (channels order is again reversed)
+    |> Enum.zip_reduce([], fn zipped_chunks, acc ->
+      [join_binaries(zipped_chunks) | acc]
+    end)
+    |> join_binaries()
+  end
```

Comment:
This seems overly complicated
1. Make sure `get_payloads` gets only a multiple of sample size.
2. Chunk the data into samples using [`Bunch.Binary.chunk_every`](https://hexdocs.pm/bunch/Bunch.Binary.html#chunk_every/2) (step 1. and 2. can be done using `Bunch.Binary.chunk_every_rem`)
3. Zip and join payloads: `Enum.zip_with(&Enum.join/1) |> Enum.join()` or `Enum.zip_with(& &1) |> Enum.flatten() |> Enum.join()` (not sure which one is better performance-wise, so choose more readable)
4. Recreate the pad map (like you do)

This probably gets rid of all the helpers in this file invalidating the comments there, but I'll leave them for educational purposes :wink: